Announcing survival analysis with tidymodels (#691)
tidyverse/tidyverse.org: content/blog/tidymodels-survival-analysis/index.Rmd

    This can be omitted if it would just recapitulate the title.

photo:
  url: https://unsplash.com/photos/n6vS3xlnsCc
  author: Kelley Bozarth

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [learn] 
tags: [tidymodels, parsnip, censored, workflows, yardstick, tune, workflowsets]
---

<!--
TODO:
* [ ] Look over / edit the post's title in the yaml
* [ ] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [ ] Find photo & update yaml metadata
* [ ] Create `thumbnail-sq.jpg`; height and width should be equal
* [ ] Create `thumbnail-wd.jpg`; width should be >5x height
* [ ] `hugodown::use_tidy_thumbnails()`
* [ ] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

We're tickled pink to announce the support of survival analysis for time-to-event data across the tidymodels framework. This makes survival analysis a first-class citizen in tidymodels and gives censored regression modeling the same flexibility and ease as classification or regression. 

> The functionality resides in mutiple tidymodels packages. The easiest way to install them all is to install the tidymodels meta-pacakge:

Reviewer: ```suggestion
The functionality resides in multiple tidymodels packages. The easiest way to install them all is to install the tidymodels meta-package:
```

```{r, eval = FALSE}
install.packages("tidymodels")
```

This blog post will highlight why this is useful, explain which additions we've made to the framework, and point to several places to learn more.

You can see a full list of changes in the release notes:

- [parsnip](https://parsnip.tidymodels.org/news/index.html#parsnip-120)
- [censored](https://censored.tidymodels.org/news/index.html#censored-030)
- [yardstick](https://yardstick.tidymodels.org/news/index.html#yardstick-130)
- [workflows](https://workflows.tidymodels.org/news/index.html#workflows-114)
- [tune](https://tune.tidymodels.org/news/index.html#tune-120)
- [finetune](https://finetune.tidymodels.org/news/index.html#finetune-120)
- [workflowsets](https://workflowsets.tidymodels.org/news/index.html#workflowsets-110)

## Increasing usefulness: Two perspectives

We'd like to situate the changes from two different perspectives: How this is useful for people already familiar with survival analysis as well as for people already familiar with tidymodels. 

If you are already familiar with both: Excellent, this is very much for you! Read on for more details on how these two things come together.

> ### Add tidymodels to your tool kit

Reviewer 2: ```suggestion
### Adding tidymodels to your tool kit
```

If you are already familiar with survival analysis but maybe not tidymodels, these changes now unlock a whole framework for predictive modelling for you. It applies tidyverse principles to modeling, meaning it strives to be consistent, composable, and human-centered. The framework covers the modeling process from the initial test/train split of the data all the way to tuning various models. Along the way it offers a rich selection of preprocessing techniques, resampling schemes, and performance metrics along with safe-guards against accidental overfitting. We make the full case for tidymodels at [tidymodels.org](https://www.tidymodels.org/).

> ### Add survival analysis to your tool kit

Reviewer 2: ```suggestion
### Adding survival analysis to your tool kit
```

> If you are already familiar with tidymodels but maybe not survival analysis, these changes let you leverage the familiar framework for an additional type of modeling problem. Survival analysis offers methods for modeling time-to-event data. While it has its roots in medical research, it has broad applications as that event of interest can be so much more than a medical outcome. Take customer churn as an example: We are interested in how long someone is a customer for and when they churn. For customers who churned, we have the complete time for which they were customers. For existing customers, we only know how long they've been customers for _so far_. Such observations are called censored. So what are our modeling choices here? 

Reviewer: Would an explicit mention of `mode = "censored regression"` in addition to `"regression"` and `'classification"` be helpful here?

> We could look at the time and model that as a regression problem. We could look at the event status and model that as a classification problem. Both options might get us somewhere close to an answer to our original modeling question but not quite there. Censored regression models let us model an outcome that includes both aspects, the time and the event status. And with that, it can deal with both censored and uncensored observations appropriately. With this type of model, we can predict the survival time, or in more applied terms, how long someone will stay a customer for overall. We can also predict the probability of survival at a given time point. This lets us answer questions like "How likely is it that this customer will churn after 3 months?". See which prediction types are available for which models at [censored.tidymodels.org](https://censored.tidymodels.org/).

Reviewer 3: I think the following sentence was a little hard to read

> how long someone will stay a customer for overall

Reviewer 3: maybe "how long someone will stay as a customer"

> We could look at the time and model that as a regression problem. We could look at the event status and model that as a classification problem. Both options might get us somewhere close to an answer to our original modeling question but not quite there. Survival analysis with its censored regression models lets us model an outcome that includes both aspects, the time and the event status. And with that, it can deal with both censored and uncensored observations appropriately. With this type of model, we can predict the survival time, or in more applied terms, how long someone will stay a customer for overall. We can also predict the probability of survival at a given time point. This let's us answer questions like "How likely is it that this customer will churn after 3 months?". See which prediction types are available for which models at [censored.tidymodels.org](https://censored.tidymodels.org/).

Reviewer: ```suggestion
We could look at the time and model that as a regression problem. We could look at the event status and model that as a classification problem. Both options might get us somewhere close to an answer to our original modeling question but not quite there. Censored regression models let us model an outcome that includes both aspects, the time and the event status. And with that, it can deal with both censored and uncensored observations appropriately. With this type of model, we can predict the survival time, or in more applied terms, how long someone will stay a customer for overall. We can also predict the probability of survival at a given time point. This lets us answer questions like "How likely is it that this customer will churn after 3 months?". See which prediction types are available for which models at [censored.tidymodels.org](https://censored.tidymodels.org/).
```

Reviewer: Generally, the phrasing of this paragraph is super helpful.

> ## Ch-ch-changes: What's new for censored regression?

Reviewer: :)

The main components needed for this full-fledged integration of survival analysis into tidymodels were

- Survival analysis models that can take censoring into account
- Survival analysis performance metrics that can take censoring into account
- Integrating changes required by these models and metrics into the framework

> For the models, parsnip gained a new mode, `"censored regression"`, for existing models as well as new model types such as `proportional_hazards()`. Engines for these reside in the parsnip extension package censored. This has been around for a while and we've previously shared [our inital thoughts](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and the [release of censored](https://www.tidyverse.org/blog/2022/08/censored-0-1-0/).

Reviewer: I tripped up on "parsnip extension package censored," though I'm not sure how we could write this clearly while adhering to tidy style.

Author: had another go at it :)

> For the models, parsnip gained a new mode, `"censored regression"`, for existing models as well as new model types such as `proportional_hazards()`. Engines for these reside in the parsnip extension package censored. This has been around for a while and we've previously shared [our inital thoughts](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and the [release of censored](https://www.tidyverse.org/blog/2022/08/censored-0-1-0/).

Reviewer: ```suggestion
For the models, parsnip gained a new mode, `"censored regression"`, for existing models as well as new model types such as `proportional_hazards()`. Engines for these reside in the parsnip extension package censored. This has been around for a while and we've previously shared [our initial thoughts](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and the [release of censored](https://www.tidyverse.org/blog/2022/08/censored-0-1-0/).
```

> For the models, parsnip gained a new mode, `"censored regression"`, for existing models as well as new model types such as `proportional_hazards()`. Engines for these reside in the parsnip extension package censored. This has been around for a while and we've previously shared [our inital thoughts](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and the [release of censored](https://www.tidyverse.org/blog/2022/08/censored-0-1-0/).

Reviewer: ```suggestion
For the models, parsnip gained a new mode, `"censored regression"`, for existing models as well as new model types such as `proportional_hazards()`. Engines for these reside in the parsnip extension package censored. The `"censored regression"` mode has been around for a while and we've previously shared posts on [our inital thoughts](https://www.tidyverse.org/blog/2021/11/survival-analysis-parsnip-adjacent/) and the [release of censored](https://www.tidyverse.org/blog/2022/08/censored-0-1-0/).
```

> Now we've added the metrics: yardstick v1.3.0 includes new metrics for assessing censored regression models. Somewhat similar to how metrics for classification models can take class predictions or probability predictions as input, these survival metrics can take predicted survival times or predictions of survival probabilities as input. 

Reviewer 3: maybe link to the changelog of the release of yardstick

The new metrics are

- Concordance index on the survival time via `concordance_survival()`
- Brier score on the survival probability and its integrated version via `brier_survival()` and `brier_survival_integrated()`

> - Area under the ROC curve on the survival probabilities via `auc_roc_survival()`

Reviewer 3: pendantic: there is also the curve function `roc_curve_survival()`

Author: Thank you, I've added it!

> The probability of survival is always defined _at a certain point in time_. We call that time point the evaluation time because it is then also the time point at which we want to evaluate model performance. Metrics that work on the survival probabilities are also called _dynamic metrics_ and you read more about them here:

Reviewer: ```suggestion
The probability of survival is always defined _at a certain point in time_. We call that time point the evaluation time because it is then also the time point at which we want to evaluate model performance. Metrics that work on the survival probabilities are also called _dynamic metrics_ and you can read more about them here:
```

> The probability of survival is always defined _at a certain point in time_. We call that time point the evaluation time because it is then also the time point at which we want to evaluate model performance. Metrics that work on the survival probabilities are also called _dynamic metrics_ and you can read more about them here:

Reviewer 2: ```suggestion
The probability of survival is always defined _at a certain point in time_. *We call that time point the evaluation time* because it is then also the time point at which we want to evaluate model performance. Metrics that work on the survival probabilities are also called _dynamic metrics_ and you can read more about them here:
```

- [Dynamic Performance Metrics for Event Time Data](https://www.tidymodels.org/learn/statistics/survival-metrics/)
- [Accounting for Censoring in Performance Metrics for Event Time Data](https://www.tidymodels.org/learn/statistics/survival-metrics-details/)

> The evaluation time is also the best example to illustrate the changes necessary to the framework. Most of them were under the hood but the evaluation time is user-facing. Let's take a look at that.

Reviewer 2: The next few paragraphs feel a little meandering. Can we make the last sentence

> Let's take a look at that:

and make the next few paragraphs bullets? Maybe that will read more focused to me.

Author: I've turned the three places of where we need to specify `eval_time` into bullet points, I think that works well.

While the need for evaluation times is dependent on type of metric, it is not actually specified as an argument to the metric functions. Like yardstick's other metrics, those take pre-made predictions as the input. So where do you specify it then?

You need to specify it to directly predict survival probabilities, via `predict()` or `augment()`. We introduced the corresponding `eval_time` argument first for fitted models in [parsnip and censored](https://www.tidyverse.org/blog/2023/04/censored-0-2-0/#introducing-eval_time) and have added it now for workflows. 

> You also need to specify it for the tuninging functions `tune_*()` from tune and finetune as they will predict survival probabilities as part of the tuning process. 

Reviewer: ```suggestion
You also need to specify it for the tuning functions `tune_*()` from tune and finetune as they will predict survival probabilities as part of the tuning process. 
```

> And lastly, the `eval_time` argument now shows up when working with tuning/resampling results such as in `show_best()` or `autoplot()`. Those changes span the packages generating and working with resampling results: tune, finetune, and workflowsets.

Reviewer: ```suggestion
Lastly, the `eval_time` argument now shows up when working with tuning/resampling results such as in `show_best()` or `autoplot()`. Those changes span the packages generating and working with resampling results: tune, finetune, and workflowsets.
```

> As we said, plenty of changes under the hood but you shouldn't need to notice them. Everything else should work "as usual", allowing the same ease and flexibility in combining tidymodels functionality for censored regression as for classification and regression.

Reviewer: ```suggestion
As we said, plenty of changes under the hood but you shouldn't need to notice them. Everything else should work "as usual," allowing the same ease and flexibility in combining tidymodels functionality for censored regression as for classification and regression.
```

This would be MLA style, but us Americans need not impose if current spelling feels cozier for you.🤪

Author: This style makes no sense to me but oh well 🏄

## The pieces come together: A case study

> To see it all in action, check out the case study [How long until building complaints are dispositioned?](https://www.tidymodels.org/learn/statistics/survival-case-study/)! 

Reviewer: ```suggestion
To see it all in action, check out the case study [How long until building complaints are dispositioned?](https://www.tidymodels.org/learn/statistics/survival-case-study/) on the tidymodels website!
```

I am pro-excitement but the punctuation marks hanging out next to each other was uncomfy for me so I added some context

Author: accidental interrobang ⁉️  😂

> The city of New York publishes data on complaints received by the Department of Buildings that include how long it takes for a complaint to be dealt with ("dispositioned") as well as several characteristics of the complaint. The case study covers a full analysis. We start with splitting the data into test and training set, explore different preprocessing strategies and model types via tuning, and predict with a final model. It should give you a good first impression of how to use tidymodels for predictive survival analysis.

Reviewer: ```suggestion
The city of New York publishes data on complaints received by the Department of Buildings that include how long it takes for a complaint to be dealt with ("dispositioned") as well as several characteristics of the complaint. The case study covers a full analysis. We start with splitting the data into test and training sets, explore different preprocessing strategies and model types via tuning, and predict with a final model. It should give you a good first impression of how to use tidymodels for predictive survival analysis.
```

Reviewer: I might propose merging this paragraph and the sentence above it?

Author: I wanted that link to be easier to spot when scanning the text quickly. I feel like it disappears too much if it's surrounded by other text on both sides, hence this split into two paragraphs.

Reviewer: Okay, you got it :)

We hope you'll find this new capability of tidymodels useful!

## Acknowledgements

> 

Reviewer 2: > Many thanks to the people who contributed to our packages since their last release:
